src/main/java/picard/analysis/CollectQualityYieldMetricsSNVQ.java

```diff
@@ -0,0 +1,548 @@
+            metricsFile.addHistogram(snvqHistogram);
+            this.collector.addHistograms(metricsFile);
+        }
+        metricsFile.write(OUTPUT);
+    }
+
+    public class QualityYieldMetricsCollector {
+        // If true, include bases from secondary alignments in metrics. Setting to true may cause double-counting
+        // of bases if there are secondary alignments in the input file.
+        private final boolean useActualBaseQualities;
+
+        // If true, include bases from secondary alignments in metrics. Setting to true may cause double-counting
+        // of bases if there are secondary alignments in the input file.
+        private final boolean includeSecondaryAlignments;
+
+        // If true, include bases from supplemental alignments in metrics. Setting to true may cause double-counting
+        // of bases if there are supplemental alignments in the input file.
+        public final boolean includeSupplementalAlignments;
+
+        // The metrics to be accumulated
+        private final QualityYieldMetrics metrics;
+
+        public QualityYieldMetricsCollector(final boolean useActualBaseQualities,
+                                            final boolean includeSecondaryAlignments,
+                                            final boolean includeSupplementalAlignments) {
+            this.useActualBaseQualities = useActualBaseQualities;
+            this.includeSecondaryAlignments = includeSecondaryAlignments;
+            this.includeSupplementalAlignments = includeSupplementalAlignments;
+            this.metrics = new QualityYieldMetrics(useActualBaseQualities);
+        }
+
+        public void acceptRecord(final SAMRecord rec, final ReferenceSequence ref) {
+            if (!this.includeSecondaryAlignments && rec.isSecondaryAlignment()) return;
+            if (!this.includeSupplementalAlignments && rec.getSupplementaryAlignmentFlag()) return;
+
+            final int length = rec.getReadLength();
+            metrics.TOTAL_READS++;
+            metrics.TOTAL_BASES += length;
+
+            final boolean isPfRead = !rec.getReadFailsVendorQualityCheckFlag();
+            if (isPfRead) {
+                metrics.PF_READS++;
+                metrics.PF_BASES += length;
+            }
+
+            // access regular quality
+            final byte[] quals;
+            if (!this.useActualBaseQualities) {
+                byte[] tmp = rec.getStringAttribute(SAMTag.BQ).getBytes();
+                if (tmp == null) {
+                    // fall back on base queslities
+                    tmp = rec.getBaseQualities();
+                } else {
+                    // zero base bq values
+                    for ( int i = 0 ; i < tmp.length ; i++ ) {
+                        tmp[i] -= 33;
+                    }
+                }
+                quals = tmp;
+            } else {
+                quals = rec.getBaseQualities();
+            }
+
+            // access snv qualities
```

Comment:
How about crashing gracefully if these tags are missing / not of the right length / etc?